src/libraries/System.Text.Json/tests/Serialization/NumberHandlingTests.cs

```diff
@@ -0,0 +1,1414 @@
+
+            JavaScriptEncoder encoder = JavaScriptEncoder.Create(encoderSettings);
+            var options = new JsonSerializerOptions(s_optionReadAndWriteFromStr)
+            {
+                Encoder = encoder
+            };
+
+            PerformEscapingTest(JsonNumberTestData.Bytes, options);
+            PerformEscapingTest(JsonNumberTestData.SBytes, options);
+            PerformEscapingTest(JsonNumberTestData.Shorts, options);
+            PerformEscapingTest(JsonNumberTestData.Ints, options);
+            PerformEscapingTest(JsonNumberTestData.Longs, options);
+            PerformEscapingTest(JsonNumberTestData.UShorts, options);
+            PerformEscapingTest(JsonNumberTestData.UInts, options);
+            PerformEscapingTest(JsonNumberTestData.ULongs, options);
+            PerformEscapingTest(JsonNumberTestData.Floats, options);
+            PerformEscapingTest(JsonNumberTestData.Doubles, options);
+            PerformEscapingTest(JsonNumberTestData.Decimals, options);
+        }
+
+        private static void PerformEscapingTest<T>(List<T> numbers, JsonSerializerOptions options)
+        {
+            // All input characters are escaped
+            IEnumerable<string> numbersAsStrings = numbers.Select(num => GetNumberAsString(num));
+            string input = JsonSerializer.Serialize(numbersAsStrings, options);
+            AssertListNumbersEscaped(input);
+
+            // Unescaping works
+            List<T> deserialized = JsonSerializer.Deserialize<List<T>>(input, options);
+            Assert.Equal(numbers.Count, deserialized.Count);
+            for (int i = 0; i < numbers.Count; i++)
+            {
+                Assert.Equal(numbers[i], deserialized[i]);
+            }
+
+            // Every number is written as a string, and custom escaping is not honored.
+            string serialized = JsonSerializer.Serialize(deserialized, options);
+            AssertListNumbersUnescaped(serialized);
+        }
+
+        private static void AssertListNumbersEscaped(string json)
+        {
+            var reader = new Utf8JsonReader(Encoding.UTF8.GetBytes(json));
+            reader.Read();
+            while (reader.Read())
+            {
+                if (reader.TokenType == JsonTokenType.EndArray)
+                {
+                    break;
+                }
+                else
+                {
+                    Assert.Equal(JsonTokenType.String, reader.TokenType);
+#if BUILDING_INBOX_LIBRARY
+                    Assert.True(reader.ValueSpan.Contains((byte)'\\'));
+#else
+                    bool foundBackSlash = false;
+                    foreach (byte val in reader.ValueSpan)
+                    {
+                        if (val == (byte)'\\')
+                        {
+                            foundBackSlash = true;
+                            break;
+                        }
+                    }
+
+                    Assert.True(foundBackSlash, "Expected escape token.");
```

Comment:
nit: You don't need this manual loop lookup. You can use `IndexOf` and check against -1.